source-container-build/app/source_build.py

```diff
@@ -391,9 +391,23 @@ def make_source_archive(
     repo_info = get_repo_info(source_dir)
     name_sha = f"{repo_info['name']}-{repo_info['last_commit_sha']}"
     output_archive = f"{source_archive_dir}/{name_sha}.tar.gz"
-    git_cmd = ["git", "archive", "--prefix", name_sha + "/", "--output", output_archive, "HEAD"]
-    log.debug("Generate source archive %r", git_cmd)
-    run(git_cmd, check=True, cwd=source_dir)
+
+    restore_cmd = ["git", "restore", "."]
+    log.debug("Restoring git repo %r", restore_cmd)
+    run(restore_cmd, check=True, cwd=source_dir)
+
+    submodule_cmd = ["git", "submodule", "update", "--init", "--recursive"]
+    log.debug("Initializing submodule(s) %r", submodule_cmd)
+    run(submodule_cmd, check=True, cwd=source_dir)
+
+    ls_cmd = ["git", "ls-files", "--recurse-submodules"]
+    log.debug("Generate source repo file list %r", ls_cmd)
+    git_process = run(ls_cmd, check=True, cwd=source_dir, capture_output=True, text=True)
+
+    tar_cmd = ["tar", "caf", output_archive, "--transform", f"s,^,{name_sha}/,", "-T-"]
```

Comment:
It's best practice to generate tarballs reproducibly; xref https://reproducible-builds.org/docs/env-variations/ and I recently did this in https://github.com/coreos/cargo-vendor-filterer/blob/0fa1892476943d1175089d89120eb694c0e3d11b/src/main.rs#L410 which note properly honors SOURCE_DATE_EPOCH but also falls back if that's not set to the toplevel git commit timestamp.

Reply:
Looks like the tar manual has guidance on this: https://www.gnu.org/software/tar/manual/html_section/Reproducibility.html

Reply:
I think that's too much to take on in this code at this point in time.

Reply:
I think setting at least the `--mtime` would be worth it, so that we get identical output when processing the same commit of a repo

```bash
$ function check_reproducibility() {
  git clone https://github.com/redhat-appstudio/build-tasks-dockerfiles /tmp/build-tasks-dockerfiles 2>/dev/null
  pushd /tmp/build-tasks-dockerfiles >/dev/null
  git ls-files | tar cza --mtime=$(git show -s --format=$cI) -T - | sha256sum
  popd >/dev/null
  rm -rf /tmp/build-tasks-dockerfiles
}

for i in {1..5}; do check_reproducibility; done
354e9e22a782f5773448355b44652b229b824936d7e1abbdedead7eb9b3330b0  -
354e9e22a782f5773448355b44652b229b824936d7e1abbdedead7eb9b3330b0  -
354e9e22a782f5773448355b44652b229b824936d7e1abbdedead7eb9b3330b0  -
354e9e22a782f5773448355b44652b229b824936d7e1abbdedead7eb9b3330b0  -
354e9e22a782f5773448355b44652b229b824936d7e1abbdedead7eb9b3330b0  -
```

Reply:
git-archive does that as well: https://git-scm.com/docs/git-archive#_description

> When a tree ID is provided, the current time is used as the modification time of each file in the archive. On the other hand, when a commit ID or tag ID is provided, the commit time as recorded in the referenced commit object is used instead.